### ISSUE TITLE
Problem using 'lein trampoline' with cljsbuild

### ISSUE THREAD

**Issue by deg** [emezeske/lein-cljsbuild#204]
There seems to be a conflict using 'lein trampoline' when my project includes ':hooks [leiningen.cljsbuild]' in the top-level map in project.clj.  Operations as simple as 'lein trampoline run' silently exit, with no error message, while 'lein run' works fine.

This bug-report is still half-baked. I'm sending what I've found so far, in the hope that someone will shout back "known bug" or "don't do that", but I can certainly give more detail if this is a real issue I've hit.

What I do know so far:
- My project is based on https://github.com/magomimmo/modern-cljs. I believe (but have not 100% tested) that everything I'm saying here applies to that project too.
- I've seen the problem in lein-cljsbuild 0.3.0 and 0.3.2. I've not tested other versions.
- When I moved the  ':hooks [leiningen.cljsbuild]' line into a :dev profile, the problem seems to go away. Even 'lein with-profile dev trampoline run' works. (Previously, I had the hook in the top-level map in my projects.clj).

**Comment by cprice404** [comment]
ran into this with @cursiveclojure today. :(

**Comment by gdanov** [comment]
me too. it seems to be one of the things getting in the way of having cursive + cljs repls play together

**Comment by conan** [comment]
This is preventing me from running REPLs in Cursive, sadly.

**Comment by cprice404** [comment]
Doing battle with this bug again today.  I think I've broken it down into two discrete issues.

The first has to do with the fact that this git repo actually produces two separate artifacts; the lein plugin artifact `lein-cljsbuild` (from the `plugin` directory), and another artifact called just `cljsbuild` (from the `support` directory).

The code in `plugin/src/leiningen/cljsbuild/subproject.clj` does some fairly magical stuff to add in a dependency on the `cljsbuild` artifact, but I don't think that's necessarily causing the problem.

I'm not sure that I understand exactly why, but the temporary script that lein generates for the trampoline execution includes this at the top:

```clj
    (require (quote cljsbuild.compiler)
             (quote cljsbuild.crossover)
             (quote cljsbuild.util))
```

And the classpath that lein launches that script with does not include any of the lein plugin artifacts, nor their dependencies, so `cljsbuild` is not on the classpath.

This is easy enough to remedy; just add `[cljsbuild "1.1.2"]` to the `dev` dependencies for your project.  Doing that gets me past the `Could not locate cljsbuild/compiler__init.class or cljsbuild/compiler.clj on classpath` error.

Which brings us to the second issue, which is the one that @ccfontes mentioned above; in the happy path of the code in `plugin/src/leiningen/cljsbuild.clj` there is a line that calls `System/exit` after the cljs compilation has completed.  When this happens it kills the trampolined repl session.

I can't find anything in the code that explains why that `exit` is necessary, and it doesn't seem like `lein cljsbuild once` runs any slower without it, so I can file a PR to remove it.

If I do the two things above (add `cljsbuild` to my `dev` deps, and remove the `exit` call from `cljsbuild.clj` in the plugin), the trampoline REPL works fine for me both from the CLI and via @cursive-ide .

**Comment by cprice404** [comment]
I filed a PR, and also deployed a fork with the fix to clojars:

https://clojars.org/cprice404/lein-cljsbuild/versions/1.1.2-noexit

**Comment by mneise** [comment]
@cprice404 Awesome, thank you for digging into this! Will have a closer look at your PR and your findings in the next couple of days :wink:

**Comment by cprice404** [comment]
@mneise ping!  just checking to see if this slipped through the cracks.

**Comment by mneise** [comment]
@cprice404 Merged your PR for the `System/exit` #436, thank you! This definitely helps when chaining tasks with `lein do`.

Regarding the missing `cljsbuild` dependency, we already add it as a dependency to the project's dependencies before running `eval-in-project`, so there really should be no need for a user to include it again in their project's dependencies. I've also noticed that not every use of `lein trampoline` has `cljsbuild` missing on the classpath, for example `lein trampoline do clean, cljsbuild once dev` works fine for me. Therefore, I think that this might be a bug in leiningen and I've already created an issue for this (https://github.com/technomancy/leiningen/issues/2095).

**Comment by cprice404** [comment]
@mneise thanks!  Adding `cljsbuild` as an explicit dependency in my dev profile is a sufficient workaround for now, but I've been digging in the lein source lately for other reasons, so maybe I'll take a peek at that bug you opened if I get a chance.

Any idea when you might do the next release?  Looking forward to getting off of my fork :)

**Comment by mneise** [comment]
@cprice404 Will try to cut a release this weekend :wink:

**Comment by cprice404** [comment]
yay, thanks.

**Comment by cprice404** [comment]
I presume this went into the 1.1.3 release?  If so, thanks!

**Comment by mneise** [comment]
@cprice404 Yes, just released it this morning. Let me know if you experience any issues :wink:

**Comment by cprice404** [comment]
will do, thanks again.

**Comment by cprice404** [comment]
Upgraded our project to 1.1.3 today and it seems to be working fine.  Thanks!

**Comment by maiq123** [comment]
Hi is there any update on this? Or are we still using the same workaround of adding the dependency?
